src/ImageSharp/Processing/Processors/Transforms/IResampler.cs

```diff
@@ -21,5 +24,54 @@ public interface IResampler
         /// The <see cref="float"/>
         /// </returns>
         float GetValue(float x);
+
+        /// <summary>
+        /// Applies an resizing transformation upon an image.
+        /// </summary>
+        /// <typeparam name="TPixel">The pixel format.</typeparam>
+        /// <param name="configuration">The configuration.</param>
+        /// <param name="source">The source image.</param>
+        /// <param name="destination">The destination image.</param>
+        /// <param name="sourceRectangle">The source bounds.</param>
+        /// <param name="targetRectangle">The target location.</param>
+        /// <param name="compand">Whether to compress or expand individual pixel color values on processing.</param>
+        void ApplyResizeTransform<TPixel>(
+            Configuration configuration,
+            Image<TPixel> source,
+            Image<TPixel> destination,
+            Rectangle sourceRectangle,
+            Rectangle targetRectangle,
+            bool compand)
+            where TPixel : struct, IPixel<TPixel>;
+
+        /// <summary>
+        /// Applies an affine transformation upon an image.
+        /// </summary>
+        /// <typeparam name="TPixel">The pixel format.</typeparam>
+        /// <param name="configuration">The configuration.</param>
+        /// <param name="source">The source image frame.</param>
+        /// <param name="destination">The destination image frame.</param>
+        /// <param name="matrix">The transform matrix.</param>
+        void ApplyAffineTransform<TPixel>(
+            Configuration configuration,
+            ImageFrame<TPixel> source,
+            ImageFrame<TPixel> destination,
+            Matrix3x2 matrix)
+            where TPixel : struct, IPixel<TPixel>;
+
+        /// <summary>
+        /// Applies a projective transformation upon an image.
+        /// </summary>
+        /// <typeparam name="TPixel">The pixel format.</typeparam>
+        /// <param name="configuration">The configuration.</param>
+        /// <param name="source">The source image frame.</param>
+        /// <param name="destination">The destination image frame.</param>
+        /// <param name="matrix">The transform matrix.</param>
+        void ApplyProjectiveTransform<TPixel>(
+            Configuration configuration,
+            ImageFrame<TPixel> source,
+            ImageFrame<TPixel> destination,
+            Matrix4x4 matrix)
```

Comment:
The only issue with these methods (and also with `IDither` ones) is that they are extending the public API surface with arbitrary stuff that I would normally consider an implementation detail. It's hard to imagine any use case for consuming them outside the ImageSharp core library.

We should investigate if we can do a trick similar to the one with `TPixel` -> `PixelOperations<TPixel>` so we can hide them from the public. We can do it in a follow-up PR though, since it's not easy to find a solution (maybe there is none).

Reply:
Yeah, this is something I tried my very best to avoid in both scenarios. 

The problem I found was with virility. Adding the generic constraints mean they had to be propagated all the way up the call stack. `.Resize()` becomes `.Resize<TResampler>()` and `ResizeOptions` becomes `ResizeOptions<TResampler>`. 

Exposing the extension classes and interface members and calling the method from inside ech implementation was my only workaround. 

Quantize was even worse. I found myself stuck requiring the `TDither` type parameter in the `IImageEncoder`. 

If there's a way around it, it's beyond my capabilities to discover.